rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -410,6 +567,9 @@ def training_states_actions_and_entities(
                 tracker, domain, omit_unset_slots=omit_unset_slots
             )
 
+            if ignore_action_unlikely_intent:
+                states = self._remove_action_unlikely_intent(states)
+
```

Comment:
Can't we move this `if` construct just before `trackers_as_states.append(states[:-1])`. That way you don't need the `if` construct on line 584 as well. Right now, the logic will be fragile if someone messes with any of these `if` constructs, hence cleaning up the actions once the states have been created is a better option IMO.

Alternatively, you can do it before `self._remove_user_text_if_intent(trackers_as_states)`. Either way, you won't need the `if` construct inside the `for` loop over tracker events.
What do you think?

Reply:
I also need the if statement in the event for loop to not include `action_unlikely_intent` events as labels (although maybe I am not fully understanding the implications of your suggestion). What if I do something like this:
- Create the event list with tracker_states
- Filter the event list in the same if statement that tracker_states is being filtered
This removes the need for the second if statement in the event for loop.

This would look like this:
```
        tracker_states = self._create_states(
            tracker, domain, omit_unset_slots=omit_unset_slots
        )
        events = tracker.applied_events()   ### <-- Create event list here

        if ignore_action_unlikely_intent:
            tracker_states = self._remove_action_unlikely_intent(tracker_states)
            events = [          ### <-- Filter events here when I filter the states
                event for event in events
                if not isinstance(event, ActionExecuted) or event.action_name != ACTION_UNLIKELY_INTENT_NAME
            ]

        label_index = 0
        entity_data = {}
        for event in events:
            if isinstance(event, UserUttered):
                entity_data = self._entity_data(event)

            elif isinstance(event, ActionExecuted):
             
                label_index += 1      ### <-- No if statement here any more, can assume we have a valid state label
                ### Do slicing and yielding here ...

```

Something similar can be done for the `IntentMaxHistoryTrackerFeaturizer`.
Additionally, per your comment below, I can move the state slicing out of this method and into `training_states_labels_and_entities` if you still think that is needed after this change.

Reply:
Hmm, that's a good point about pruning them for labels. I agree with your suggestion above 👍 
Just a minor (maybe it was just a typo from you) - it will be `if not isinstance(event, ActionExecuted) and event.action_name != ACTION_UNLIKELY_INTENT_NAME` and not `if not isinstance(event, ActionExecuted) or event.action_name != ACTION_UNLIKELY_INTENT_NAME`

Reply:
State slicing can stay there only then.